File: go/projections/event_metadata.go
package projections

import (
	"github.com/fraym/freym-api/go/proto/projections/deliverypb"
)

type EventMetadata struct {
	CorrelationId string
	CausationId   string
	Target        deliverypb.DeploymentTarget
}

func (m *EventMetadata) getProtobufEventMetadata() *deliverypb.EventMetadata {
	if m == nil {
		return nil
	}

	return deliverypb.EventMetadata_builder{
		CorrelationId: m.CorrelationId,
		CausationId:   m.CausationId,
		Target:        m.Target,
	}.Build()
}

File: js/projections/client/eventMetadata.ts
import {
    DeploymentTarget,
    EventMetadata as ProjectionsEventMetadata,
} from "@fraym/proto/dist/index.freym.projections.delivery";

export interface EventMetadata {
    causationId: string;
    correlationId: string;
}

export const fillMetadataWithDefaults = (
    metadata: Partial<EventMetadata> | null,
    target: DeploymentTarget
): ProjectionsEventMetadata => {
    if (!metadata) {
        return {
            causationId: "",
            correlationId: "",
            target,
        };
    }

    return {
        causationId: metadata.causationId || "",
        correlationId: metadata.correlationId || "",
        target,
    };
};
